### DISCUSSION TITLE
인메모리 데이터 베이스는 어떤 특징이 있나요?

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#133] [Database]
H2를 쓰기도 했으니까 질문해봅니다

**Comment by jimi567** [comment]
인 메모리 DB는 데이터 저장에 내부 메모리를 주로 사용하는 목적별 데이터베이스 이다.(h2, Redis 등이 있음)
기존 DB(Mysql, MongoDB 등...) 가 디스크에 데이터를 저장하는 반면, 메모리에 데이터를 업로드하기 때문에 다음과 같은 장점을 누릴 수 있다.

* 짧은 지연 시간 으로 실시간 응답 제공 가능  (지연시간? 데이터 액세스 요청과 애플리케이션 응답 사이에 발생하는 지연)

* 높은 처리량  (처리량 ? 특정 기간 동안의 읽기, 쓰기 작업 수)

* 높은 확장성 : 성능에 악영향을 미치지 않으면서 쓰기 크기와 읽기 크기를 모두 조정할 수 있음


--- 
번외 - 일반 DB와 비교

|     |디스크 기반 DB|인 메모리 DB|
|------|---|---|
|저장장치|디스크 중심|메모리 중심|
|인덱스|B-tree|T-tree|
|DB 위치|디스크|메모리|
|DB크기|디스크 크기|가상 메모리 크기|
|사용 목적|안정성/대용량 데이터 처리|빠른 트랜잭션 처리 / 빠른 응답|
|회복기법|UNDO/REDO/CheckPoint/Shadow Paging|보조 메모리/ 디스크  (메모리에 존재하여 이중화 형태로 제공)|

**Comment by GIVEN53** [comment]
인메모리 데이터베이스(In-Memory Database, IMDB)는 모든 데이터를 메인 메모리(RAM)에 저장하고 관리하는 데이터베이스 시스템입니다. 전통적인 디스크 기반 데이터베이스와는 달리, 인메모리 데이터베이스는 데이터 접근 속도와 성능이 뛰어납니다.

### 주요 특징
- 고속 성능
메모리에 데이터를 저장하므로 디스크 I/O가 필요 없어서 데이터 접근 속도가 매우 빠릅니다. 이는 특히 읽기 및 쓰기 작업에서 큰 성능 향상을 가져옵니다.
- 낮은 지연 시간
데이터 접근과 처리 시간이 짧기 때문에 지연 시간이 적습니다. 실시간 분석 및 고성능 애플리케이션에서 중요한 역할을 합니다.
- 단순한 구조
데이터가 메모리에 저장되므로 데이터 관리가 단순합니다. 디스크 기반 데이터베이스처럼 복잡한 캐싱 메커니즘이나 I/O 최적화가 필요 없습니다.
- 데이터 내구성
메모리는 휘발성이므로 전원이 꺼지면 데이터가 사라집니다. 이를 보완하기 위해 인메모리 데이터베이스는 정기적으로 데이터를 디스크에 백업하거나, 복제 기능을 사용하여 데이터 내구성을 확보합니다.

### 인메모리 데이터베이스의 장단점
#### 장점
- 속도: 디스크 I/O를 제거하여 매우 빠른 데이터 접근 속도 제공.
- 실시간 처리: 실시간 데이터 분석 및 처리가 가능하여 빠른 응답이 요구되는 애플리케이션에 적합.
- 단순성: 복잡한 캐싱 메커니즘이 필요 없으며, 데이터 관리가 용이함.
#### 단점
- 데이터 휘발성: 전원이 꺼지면 메모리에 저장된 데이터가 사라질 수 있음. 이를 보완하기 위해 추가적인 복제나 백업 메커니즘이 필요.
- 비용: 대용량 메모리는 디스크보다 비용이 많이 들기 때문에 경제적으로 부담이 될 수 있음.
- 메모리 제약: 메모리 크기에 따라 저장할 수 있는 데이터의 양이 제한됨.

**Comment by 3Juhwan** [comment]
인메모리 데이터베이스
- 빠른 데이터 접근 속도, 낮은 지연 시간
  - 메모리 기반
  - 디스크 I/O 없음
- 단순함
- 데이터 휘발성
  - 복제, 백업
- 비싼 자원
- 메모리 크기에 한계
